File: composeApp/src/androidMain/kotlin/co/jarias/flexapp/ui/screens/welcome/WelcomeScreenViewModel.kt
package co.jarias.flexapp.ui.screens.welcome

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.jarias.flexapp.data.local.PreferencesManager
import co.jarias.flexapp.data.local.ToolType
import co.jarias.flexapp.ui.navigation.NavigationEvent
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class WelcomeScreenViewModel(
    private val preferencesManager: PreferencesManager? = null
) : ViewModel() {

    private val _state = MutableStateFlow(WelcomeScreenState(isLoading = true))
    val state: StateFlow<WelcomeScreenState> = _state.asStateFlow()

    init {
        loadLastTool()
    }

    private fun loadLastTool() {
        viewModelScope.launch {
            val lastTool = preferencesManager?.getLastTool()
            _state.value = _state.value.copy(
                isLoading = false,
                lastTool = lastTool
            )
        }
    }

    fun onEvent(event: WelcomeScreenEvents, onNavigate: (NavigationEvent) -> Unit) {
        when (event) {
            is WelcomeScreenEvents.OnGetStartedClicked -> {
                val lastTool = _state.value.lastTool
                if (lastTool != null) {
                    onNavigate(NavigationEvent.NavigateToTool(lastTool))
                } else {
                    onNavigate(NavigationEvent.NavigateToToolSelection)
                }
            }
        }
    }
}

File: iosApp/iosApp/Screens/welcome/WelcomeScreenViewModel.swift
import Foundation
import shared

class WelcomeScreenViewModel: ObservableObject {
    @Published var state = WelcomeScreenState()
    private let getLastToolUseCase: GetLastToolUseCase
    
    init() {
        self.getLastToolUseCase = KoinHelper().getLastToolUseCase
        loadLastTool()
    }
    
    private func loadLastTool() {
        getLastToolUseCase.invoke { [weak self] tool, error in
            DispatchQueue.main.async {
                self?.state.lastTool = tool
                self?.state.isLoading = false
            }
        }
    }
    
    func onEvent(_ event: WelcomeScreenEvents, onNavigate: (NavigationEvent) -> Void) {
        switch event {
        case .onGetStartedClicked:
            if let lastTool = state.lastTool {
                onNavigate(.navigateToTool(toolType: lastTool))
            } else {
                onNavigate(.navigateToToolSelection)
            }
        }
    }
}
